#!/bin/bash
#
#SBATCH --job-name=PINFISH_R
#SBATCH --error=AVAS_gene_modeling_pinfish_run.log
#SBATCH --nodes=1
#SBATCH --ntasks=32
#SBATCH --mem=60G
#SBATCH --qos=low
#SBATCH --mail-user=m.eitel@lmu.de
#SBATCH --mail-type=END,FAIL


###############
### PRESETS ###
###############



NUMTH=32
REF=Aphrocallistes_instagraal_l4_n50_c1_N5.polished_sorted_hardmasked.fasta
TEMP=/home/meitel/data/avas/pinfish/scaffolded_HiC/lordec-corrected/tmp



#############################
### PINFISH gene modeling ###
#############################


source activate pinfish

## The mapped reads need to be previously filtered for unmapped and secondary alignments.
# spliced_bam2gff converts sorted BAM files containing spliced alignments (GMAP) into GFF2 format.
spliced_bam2gff -t $NUMTH ${REF%.fasta}.ONT-RNA_minimap2_mapped_singletons_filtered_sorted.bam \
> PINFISH_stepwise_corrected-reads_raw_transcripts.gff \
2> PINFISH_spliced_bam2gff.log

# cluster_gff takes the sorted GFF2 file as input and clusters together reads having similar exon/intron structure to create a rough consensus of the clusters by taking the median of exon boundaries from all transcripts in the cluster.
cluster_gff -t $NUMTH -a PINFISH_stepwise_corrected-reads_clustered_transcripts.tsv \
PINFISH_stepwise_corrected-reads_raw_transcripts.gff \
> PINFISH_stepwise_corrected-reads_clustered_transcripts.gff \
2> PINFISH_cluster_gff.log

## Polish cluster is the step that takes the longest and it will fail if the bam file is not properly filtered and sorted.
# polish_clusters takes the cluster definitions generated by cluster_gff and creates an error corrected read for each cluster by mapping all reads on the read with the median length (using GMAP) and polishing it using racon. The polished reads can be mapped to the genome using MINIMAP2 or GMAP.
polish_clusters -t $NUMTH -d $TEMP \
-a PINFISH_stepwise_corrected-reads_clustered_transcripts.tsv \
-o PINFISH_stepwise_corrected-reads_clustered_transcripts.fasta \
${REF%.fasta}.ONT-RNA_minimap2_mapped_singletons_filtered_sorted.bam \
2> PINFISH_polish_clusters.log

# collapse_partials takes GFFs generated by either cluster_gff or polish_clusters and filters out transcripts which are likely to be based on RNA degradation products from the 5' end. The tool clusters the input transcripts into "loci" by the 3' ends and discards transcripts which have a compatible transcripts in the loci with more exons.
collapse_partials -t $NUMTH PINFISH_stepwise_corrected-reads_clustered_transcripts.gff \
> PINFISH_stepwise_corrected-reads_clustered_transcripts_collapsed.gff \
2> PINFISH_collapse_partials.log


conda deactivate

###################################
### TRANSDECODER $ REFormatting ###
###################################


source activate mappings

gffread -T PINFISH_stepwise_corrected-reads_clustered_transcripts_collapsed.gff > PINFISH_stepwise_corrected-reads_clustered_transcripts_collapsed.gtf

# extract STRINGTIE transcripts and generate a gff3 file:
gtf_genome_to_cdna_fasta.pl PINFISH_stepwise_corrected-reads_clustered_transcripts_collapsed.gtf $REF > PINFISH_stepwise_corrected-reads_clustered_transcripts_collapsed_mRNA.fasta
gtf_to_alignment_gff3.pl PINFISH_stepwise_corrected-reads_clustered_transcripts_collapsed.gtf > PINFISH_stepwise_corrected-reads_clustered_transcripts_collapsed.gff3

# predict coding sequences with TRANSDECODER v.2.1.0:
TransDecoder.LongOrfs -t PINFISH_stepwise_corrected-reads_clustered_transcripts_collapsed_mRNA.fasta
TransDecoder.Predict -t PINFISH_stepwise_corrected-reads_clustered_transcripts_collapsed_mRNA.fasta


#mv and rename transdecoder outputs:
mv PINFISH_stepwise_corrected-reads_clustered_transcripts_collapsed_mRNA.fasta.transdecoder.cds PINFISH_stepwise_corrected-reads_clustered_transcripts_collapsed_transdecoder_CDS.fasta
mv PINFISH_stepwise_corrected-reads_clustered_transcripts_collapsed_mRNA.fasta.transdecoder.gff3 PINFISH_stepwise_corrected-reads_clustered_transcripts_collapsed_transdecoder_mRNA.gff3
mv PINFISH_stepwise_corrected-reads_clustered_transcripts_collapsed_mRNA.fasta.transdecoder.pep PINFISH_stepwise_corrected-reads_clustered_transcripts_collapsed_transdecoder_proteins.fasta


# generate STRINGTIE transdecoder CDS .gff3 file with GMAP:
gmap_build -t $NUMTH -d ${REF%.fasta} $REF
gmap -d ${REF%.fasta} PINFISH_stepwise_corrected-reads_clustered_transcripts_collapsed_transdecoder_CDS.fasta -f 3 -B 4 -t $NUMTH -n 1 > PINFISH_stepwise_corrected-reads_clustered_transcripts_collapsed_transdecoder_CDS.gff3 2> PINFISH_stepwise_corrected-reads_clustered_transcripts_collapsed_transdecoder_CDS.log

# generate STRINGTIE transdecoder mRNA .gff3 file with GMAP:
gmap -d ${REF%.fasta} PINFISH_stepwise_corrected-reads_clustered_transcripts_collapsed_mRNA.fasta -f 3 -B 4 -t $NUMTH -n 1 > PINFISH_stepwise_corrected-reads_clustered_transcripts_collapsed_mRNA.gff3 2> PINFISH_stepwise_corrected-reads_clustered_transcripts_collapsed_mRNA.log


conda deactivate


